lib/matplotlib/axes/_base.py

```diff
@@ -3397,6 +3397,11 @@ def get_xticklabels(self, minor=False, which=None):
         -------
         ret : list
            List of `~matplotlib.text.Text` instances.
+        
+        Notes
+        -----
+        Please note that the tick label list is not populated until ``plt.draw()`` has been called.
+        See also: https://stackoverflow.com/questions/32700935/get-xticklabels-contains-empty-text-instances
```

Comment:
I dont' think we usually call out to outside resources (they are fragile).  If there is something pertinent in that post please include here.  

I think this is actually a bug, and the lazy tick instantiation should be triggered by `get_xticklabels`.  I think before we did this lazily, this used to work?  ping @anntzer @efiring

Reply:
I raised an issue for this (#16140). I'd be happy to expand on that and provide a minimal example if it helps.

Reply:
I think we never did (at least AFAIK) and the current behavior was always present, but I agree we could consider changing it.

I also agree the SO link could be skipped.

Reply:
I cannot remember of this ever being different than it is now. It's pretty inherent to the fact that the locator first needs to determine the ticks and the formatter finding their labels. This can only be done when drawing the axes - else, it might still change afterwards.  
For that reason I would actually argue for not changing the behaviour; because as it stands, we can at least be sure that `get_xticklabels()` will never return anything that is only seemingly correct.

Adding a note in the docs seems useful though. (No SO link needed IMO. Instead a reference to `pyplot.draw` and `canvas.draw`)

Also, I would dare to claim that in 90% of the cases where people might want to use `.get_{x,y}ticklabels`, there would be a better solution. (E.g. in the linked SO post, they want to rotate labels, for which one does not actually need to know the ticklabels.)

Reply:
Well, then I think this leads to the question: what is the point of this function at all?  Presumably its to get the tick labels and change them (i.e. add a dollar sign or a happy face to all of them).  If the user wants to do this, and avoid writing their own formatter, then I'd argue they shouldn't need to call an explicit `draw` in order to do so.  Or, if we do need to call a draw (because it's impossible to know the limits before a draw) then maybe this function should call the `draw`.  But I think the function is fundamentally broken if we have to explicitly tell users to call `draw` to get it to work.

Reply:
I don't think the function is fundamentally broken. It returns a list of the `Text` instances of the labels of the current state of the axes, just as it should. 
Maybe the problem is more what users expect from it - and that could indeed be clarified in the documentation.

I think it's similar to how `ax.get_position()` returns the current position, and requires a `draw` for it to be the actual position in the final figure.

Reply:
Agree not to include the SO link (we are the canonical reference, SO should link to _us_ ;) ).

Agree with @ImportanceOfBeingErnest that this function is fine as it is.  Putting `draws` in can get very expensive very fast and you can get fun re-entrant behaviour if it were to be called from _inside_ of a draw.

We use this function a bit internally and a whole lot in the examples, (`plt.setp(ax.get_xticklabel(), ...)`)

Reply:
I'm not following the analogy to `ax.get_position`; that doesn't require a draw to return the current position.  The only thing that I know of that requires a draw is `text.get_extent` and thats because we don't render the text until draw time, and hence have no way to know the extent.  

I'm not suggesting we need to do a whole draw, but it seems populating the ticks is not an unreasonable thing to do if the user asks for them.   I'm actually not following how our internal uses of `get_xticklabels` even work if the number of ticks change (i.e.  `autofmt_xdate`)

Reply:
At least before 3.0 `ax.get_position()` used to give wrong positions unless you drew the canvas. Not sure what exactly changed in between. But to construct an example where this is still the case, we can use an inset axes:

```
import matplotlib.pyplot as plt
import numpy as np

fig, ax = plt.subplots()
ax.imshow(np.random.rand(10,4))
axins = ax.inset_axes([0.0, 0.4, 0.7, 0.5])

print(axins.get_xticklabels())
# prints "wrong" labels: <a list of 6 Text xticklabel objects>
print(axins.get_position())
# prints "wrong" position: Bbox(x0=0.125, y0=0.418, x1=0.6675, y1=0.803)

fig.canvas.draw()

print(axins.get_xticklabels())
# prints "correct" labels: <a list of 3 Text xticklabel objects>
print(axins.get_position())
# prints "correct" position: Bbox(x0=0.397, y0=0.418, x1=0.5587, y1=0.803)
```

Reply:
OK, but thats a pretty specialized example using `set_axes_locator`.  I agree that any axes that use `set_axes_locator (including everything in axes_grid) will not return a sensible position until the locator function is called.

Reply:
... though come to think of it, its possible the axes_locator function could be called if `get_position()` is called.

Reply:
Hi everyone,

I've updated the documentation improvement suggestion:
* removed the link to SO
* added references to `pyplot.draw` and `canvas.draw`

I guess the same problem will also apply to the `yticklabels` method? A colleague and I had the idea that it might if the function should not be changed, maybe it would at least be helpful to add a warning message so the user is informed that the tick label list has not yet been populated. Would something like this be feasible?

Reply:
Yes, `set_yticklabels` should get the same text. :+1: 

I'm -0.5 on adding a warning. We tend to be cautious to issue warnings when we are not sure something is going wrong. It's annoying to issue warnings for valid use cases. There are (few) cases when one might want to have the tick `Text` instances and the actual string values are irrelevant.